timemachine/md/smc.py

```diff
@@ -0,0 +1,190 @@
+    resample: function
+        (optionally) perform resampling given an array of log weights
+
+    Returns
+    -------
+    trajs_dict
+        "sample_traj"
+            [K-1, N] list of snapshots
+        "incremental_log_weights_traj"
+            [K-1, N] array of incremental log weights
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+
+    References
+    ----------
+    * Arnaud Doucet's annotated bibliography of SMC
+        https://www.stats.ox.ac.uk/~doucet/smc_resources.html
+    * [Dai, Heng, Jacob, Whiteley, 2020] An invitation to sequential Monte Carlo samplers
+        https://arxiv.org/abs/2007.11936
+
+    See Also
+    --------
+    * get_endstate_samples_from_smc_result
+    """
+    n = len(samples)
+    log_weights = np.zeros(n)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        sample_traj.append(samples)
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+
+    # Note: loop bounds [1:-1] intentional:
+    #   these steps are the only ones that contribute to the SMC result,
+    #   and some care is needed to extract endstate samples at t=0, t=T-1
+    #   See also
+    #   * discussion at https://github.com/proteneer/timemachine/pull/718#discussion_r854276326
+    #   * helper function get_endstate_samples_from_smc_result
+    for (lam_initial, lam_target) in zip(lambdas[:-2], lambdas[1:-1]):
+        # update log weights
+        incremental_log_weights = log_prob(sample_traj[-1], lam_target) - log_prob(sample_traj[-1], lam_initial)
+        log_weights += incremental_log_weights
+
+        # resample
+        indices, log_weights = resample(log_weights)
+        resampled = [sample_traj[-1][i] for i in indices]
+
+        # propagate
+        samples = propagate(resampled, lam_target)
+
+        # log
+        accumulate_results(samples, indices, log_weights, incremental_log_weights)
+
+    # final result: a collection of samples, with associated log weights
+    incremental_log_weights = log_prob(samples, lambdas[-1]) - log_prob(samples, lambdas[-2])
+    incremental_log_weights_traj.append(incremental_log_weights)
+    log_weights_traj.append(np.array(log_weights + incremental_log_weights))
+
+    # cast everything (except samples list) to arrays
+    trajs_dict = dict(
+        traj=sample_traj,
+        log_weights_traj=np.array(log_weights_traj),
+        ancestry_traj=np.array(ancestry_traj),
+        incremental_log_weights_traj=np.array(incremental_log_weights_traj),
+    )
+    return trajs_dict
+
+
+def null_resample(log_weights):
```

Comment:
nit: something like `identity_resample` makes slightly more sense to me here

Reply:
Applied in https://github.com/proteneer/timemachine/pull/718/commits/36eb30d93f0d70df8daa23e33b129d69956c1304